File: Javascript/106. Construct Binary Tree from Inorder and Postorder Traversal.js
/*
Given two integer arrays inorder and postorder where 
    inorder is the inorder traversal of a binary tree and 
    postorder is the postorder traversal of the same tree, construct and return the binary tree.
    Inorder Traversal → [Left, Root, Right]
    Postorder Traversal → [Left, Right, Root]
*/
function TreeNode(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
}
function printPostorder(node) {
    if (node === null) { return; }
    console.log(node.val);
    printPostorder(node.left);
    printPostorder(node.right);
}    
/**
 * 106. Construct Binary Tree from Inorder and Postorder Traversal
 * @param {number[]} inorder
 * @param {number[]} postorder
 * @return {TreeNode}
 */
var buildTree = function(inorder, postorder) {
    const n = postorder.length;
    if (!inorder || !postorder) { return null; }
    const inorderIndexMap = new Map(); 
    
    for (let i = 0; i < n; i++) {
        inorderIndexMap.set(inorder[i], i);
    }
    
    let postIndex = n - 1;
    
    const constructTree = function (inLeft, inRight) {
        if (postIndex < 0 || inLeft > inRight) { return null; }
        // Root is always the last element of postorder
        const rootVal = postorder[postIndex--];
        const root = new TreeNode(rootVal);
        // Find root index in inorder using hashmap
        let rootIndexInOrder = inorderIndexMap.get(rootVal);
        let leftSubTreeSize = rootIndexInOrder - inLeft;
        // Recursively build right subtree first (since postorder is Left-Right-Root)
        root.right = constructTree(rootIndexInOrder + 1, inRight);
        root.left = constructTree(inLeft, rootIndexInOrder - 1);
                
        return root;
    }
    
    return constructTree(0, n - 1);
};

const inorder = [9,3,15,20,7];
const postorder = [9,15,7,20,3];
// Expected [3,9,20,null,null,15,7]

printPostorder(buildTree(inorder, postorder));

File: Rust/106_Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal.rs
/*
Given two integer arrays inorder and postorder where 
    inorder is the inorder traversal of a binary tree and 
    postorder is the postorder traversal of the same tree, construct and return the binary tree.
    Inorder Traversal → [Left, Root, Right]
    Postorder Traversal → [Left, Right, Root]
*/
/* 106. Construct Binary Tree from Inorder and Postorder Traversal */
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::HashMap;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode { val, left: None, right: None }
    }
}

struct Solution;
impl Solution {
    pub fn build_tree(inorder: Vec<i32>, postorder: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        if inorder.len() != postorder.len() {
            return None; // Avoids mismatched lengths
        }
        
        fn helper(
            in_left: isize, 
            in_right: isize, 
            post_idx: &mut isize, 
            inorder_map: &HashMap<i32, usize>, 
            postorder: &Vec<i32>
        ) -> Option<Rc<RefCell<TreeNode>>> {

            if in_left > in_right || *post_idx < 0 {
                return None;
            }
            
            let root_val = postorder[*post_idx as usize];
            let root = Rc::new(RefCell::new(TreeNode::new(root_val)));
            
            let index = *inorder_map.get(&root_val).unwrap() as isize;
            *post_idx -= 1;
            
            
            root.borrow_mut().right = helper(index + 1, in_right, post_idx, inorder_map, postorder);
            root.borrow_mut().left = helper(in_left, index - 1, post_idx, inorder_map, postorder);
            
            Some(root)
        }
        
        let inorder_map: HashMap<i32, usize> = inorder.iter().enumerate().map(|(i, &val)| (val, i)).collect();
        let mut post_idx = (postorder.len() as isize) - 1;
        
        helper(0, inorder.len().saturating_sub(1) as isize, &mut post_idx, &inorder_map, &postorder)
    }
}

fn main() {
    let inorder = vec![9,3,15,20,7];
    let postorder = vec![9,15,7,20,3];
    let root = Solution::build_tree(inorder, postorder);
    println!("{:?}", root);
    // Expect [3,9,20,null,null,15,7]
}